iothub/device/samples/convention-based-samples/Thermostat/ThermostatSample.cs

```diff
@@ -0,0 +1,191 @@
+        {
+            switch (commandRequest.CommandName)
+            {
+                case "getMaxMinReport":
+                    try
+                    {
+                        DateTimeOffset sinceInUtc = commandRequest.GetData<DateTimeOffset>();
+                        _logger.LogDebug($"Command: Received - Generating max, min and avg temperature report since " +
+                            $"{sinceInUtc.LocalDateTime}.");
+
+                        Dictionary<DateTimeOffset, double> filteredReadings = _temperatureReadingsDateTimeOffset
+                            .Where(i => i.Key > sinceInUtc)
+                            .ToDictionary(i => i.Key, i => i.Value);
+
+                        if (filteredReadings != null && filteredReadings.Any())
+                        {
+                            var report = new TemperatureReport
+                            {
+                                MaximumTemperature = filteredReadings.Values.Max<double>(),
+                                MinimumTemperature = filteredReadings.Values.Min<double>(),
+                                AverageTemperature = filteredReadings.Values.Average(),
+                                StartTime = filteredReadings.Keys.Min(),
+                                EndTime = filteredReadings.Keys.Max(),
+                            };
+
+                            _logger.LogDebug($"Command: MaxMinReport since {sinceInUtc.LocalDateTime}:" +
+                                $" maxTemp={report.MaximumTemperature}, minTemp={report.MinimumTemperature}, avgTemp={report.AverageTemperature}, " +
+                                $"startTime={report.StartTime.LocalDateTime}, endTime={report.EndTime.LocalDateTime}");
+
+                            return Task.FromResult(new CommandResponse(report, StatusCodes.OK));
+                        }
+
+                        _logger.LogDebug($"Command: No relevant readings found since {sinceInUtc.LocalDateTime}, cannot generate any report.");
+
+                        return Task.FromResult(new CommandResponse(StatusCodes.NotFound));
+                    }
+                    catch (JsonReaderException ex)
+                    {
+                        _logger.LogError($"Command input for {commandRequest.CommandName} is invalid: {ex.Message}.");
+
+                        return Task.FromResult(new CommandResponse(StatusCodes.BadRequest));
+                    }
+
+                default:
+                    _logger.LogWarning($"Received a command request that isn't" +
+                        $" implemented - command name = {commandRequest.CommandName}");
+
+                    return Task.FromResult(new CommandResponse(StatusCodes.NotFound));
+            }
+        }
+
+        // Send temperature updates over telemetry. The sample also sends the value of max temperature since last reboot over reported property update.
+        private async Task SendTemperatureAsync()
```

Comment:
Why not make the main function call SendTemperatureTelemetryAsync and ReportMaxTemperatureSinceLastRebootAsync separately? The reason I felt that would be more readable was because they are both showing different APIs in the SDK and felt like they would be explicitly called out that way.

Reply:
the reason the main function doesn't call them directly is because UpdateMaxTemperatureSinceLastRebootPropertyAsync isn't always called, it is called only based on some calculation, and it seemed cleaner to abstract away the calculation logic to a different method. I've added the comment to the main func, which should help highlight the 2 calls happening in here.